### DISCUSSION TITLE
Sudden weird error when using SDXL refiner

### DISCUSSION THREAD

**Discussion by Silanda** [vladmandic/sdnext#2026] [Q&A]
Everything seemed to be working fine earlier, but I've just tried to do some gens and I'm getting this error:

13:59:29-212073 ERROR    Exception: Input type (torch.cuda.HalfTensor) and weight type (torch.HalfTensor) should be the
                         same
13:59:29-222053 ERROR    Arguments: args=('task(nbqkku8nr7cyklp)', '', '', [], 50, 1, 1, True, False, False, 1, 1, 11,
                         11, 0.7, 1, -1.0, -1.0, 0, 0, 0, 768, 1344, True, 0.3, 2, 'None', 50, 0, 0, 0.8, '', '', [], 0,
                         False, 'x264', 'blend', 10, 0, 0, False, True, True, True, 'intermediate', 'animation', False,
                         False, 'positive', 'comma', 0, False, False, '', 0, '', [], 0, '', [], 0, '', [], True, False,
                         False, False, 0, False, False, 4.0, '', 10.0, 'Linear', 3, False, 30.0, True, False, False, 0,
                         0.0, 'Lanczos', 1, True, 0, 0, 0.001, 75, 0.0, False, True) kwargs={}
13:59:29-224084 ERROR    gradio call: RuntimeError
╭───────────────────────────────────────── Traceback (most recent call last) ──────────────────────────────────────────╮
│ I:\GitHub\automatic\modules\call_queue.py:34 in f                                                                    │
│                                                                                                                      │
│    33 │   │   │   try:                                                                                               │
│ ❱  34 │   │   │   │   res = func(*args, **kwargs)                                                                    │
│    35 │   │   │   │   progress.record_results(id_task, res)                                                          │
│                                                                                                                      │
│ I:\GitHub\automatic\modules\txt2img.py:65 in txt2img                                                                 │
│                                                                                                                      │
│   64 │   if processed is None:                                                                                       │
│ ❱ 65 │   │   processed = processing.process_images(p)                                                                │
│   66 │   p.close()                                                                                                   │
│                                                                                                                      │
│ I:\GitHub\automatic\modules\processing.py:565 in process_images                                                      │
│                                                                                                                      │
│    564 │   │   else:                                                                                                 │
│ ❱  565 │   │   │   res = process_images_inner(p)                                                                     │
│    566 │   finally:                                                                                                  │
│                                                                                                                      │
│ I:\GitHub\automatic\modules\processing.py:709 in process_images_inner                                                │
│                                                                                                                      │
│    708 │   │   │   │   from modules.processing_diffusers import process_diffusers                                    │
│ ❱  709 │   │   │   │   x_samples_ddim = process_diffusers(p, p.seeds, p.prompts, p.negative_pro                      │
│    710                                                                                                               │
│                                                                                                                      │
│ I:\GitHub\automatic\modules\processing_diffusers.py:304 in process_diffusers                                         │
│                                                                                                                      │
│   303 │   │   │   if not shared.state.interrupted and not shared.state.skipped:                                      │
│ ❱ 304 │   │   │   │   refiner_images = vae_decode(latents=refiner_output.images, model=shared.                       │
│   305 │   │   │   │   results.append(refiner_images[0])                                                              │
│                                                                                                                      │
│                                               ... 4 frames hidden ...                                                │
│                                                                                                                      │
│ I:\GitHub\automatic\venv\lib\site-packages\diffusers\models\autoencoder_kl.py:256 in _decode                         │
│                                                                                                                      │
│   255 │   │                                                                                                          │
│ ❱ 256 │   │   z = self.post_quant_conv(z)                                                                            │
│   257 │   │   dec = self.decoder(z)                                                                                  │
│                                                                                                                      │
│ I:\GitHub\automatic\venv\lib\site-packages\torch\nn\modules\module.py:1501 in _call_impl                             │
│                                                                                                                      │
│   1500 │   │   │   │   or _global_forward_hooks or _global_forward_pre_hooks):                                       │
│ ❱ 1501 │   │   │   return forward_call(*args, **kwargs)                                                              │
│   1502 │   │   # Do not call functions when jit is used                                                              │
│                                                                                                                      │
│ I:\GitHub\automatic\extensions-builtin\Lora\lora.py:417 in lora_Conv2d_forward                                       │
│                                                                                                                      │
│   416 │                                                                                                              │
│ ❱ 417 │   return torch.nn.Conv2d_forward_before_lora(self, input)                                                    │
│   418                                                                                                                │
│                                                                                                                      │
│ I:\GitHub\automatic\venv\lib\site-packages\torch\nn\modules\conv.py:463 in forward                                   │
│                                                                                                                      │
│    462 │   def forward(self, input: Tensor) -> Tensor:                                                               │
│ ❱  463 │   │   return self._conv_forward(input, self.weight, self.bias)                                              │
│    464                                                                                                               │
│                                                                                                                      │
│ I:\GitHub\automatic\venv\lib\site-packages\torch\nn\modules\conv.py:459 in _conv_forward                             │
│                                                                                                                      │
│    458 │   │   │   │   │   │   │   _pair(0), self.dilation, self.groups)                                             │
│ ❱  459 │   │   return F.conv2d(input, weight, bias, self.stride,                                                     │
│    460 │   │   │   │   │   │   self.padding, self.dilation, self.groups)                                             │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
RuntimeError: Input type (torch.cuda.HalfTensor) and weight type (torch.HalfTensor) should be the same

Anyone got any ideas?

EDIT: I should probably add that this seems to happen when running the refiner.

**Comment by Silanda** [comment]
It looks like the "Move base model to CPU when using refiner" option has broken. Without it, it generates without error but extremely slowly (or just slowly if "Enable sequential CPU offload") is enabled.

Enabling "Enable model CPU offload" also seems to fix the issue, though there seems to be a little performance loss.

**Comment by vladmandic** [comment]
i need to make code perform some sanity checks to disallow usage of both - basically, you can use either settings from first section:
```python
    "diffusers_move_base": OptionInfo(False, "Move base model to CPU when using refiner"),
    "diffusers_move_unet": OptionInfo(False, "Move base model to CPU when using VAE"),
    "diffusers_move_refiner": OptionInfo(True, "Move refiner model to CPU when not in use"),
```
OR from second section - but you cannot mix them.
```python
    "diffusers_seq_cpu_offload": OptionInfo(False, "Enable sequential CPU offload"),
    "diffusers_model_cpu_offload": OptionInfo(False, "Enable model CPU offload"),
```

those params perform similar ops, difference is that first section is implemented natively in sdnext and second section is built-in diffusers functionality. so if you use one from first section and one from second, no clue where is the model offloaded to - its bound to break.

**Comment by Silanda** [comment]
Thing is, I wasn't mixing them originally; I was only using the options from the first section. I only tried the second section to see if it made any difference. It does, in that it works, though the result is exactly the same as if options from the second section are applied alone.

The "Move base model to CPU when using refiner" option has been broken for me and others (judging from the Discord) since yesterday.

EDIT: To be absolutely clear: no options from the second section were enabled when the error in the first post was produced. That error will occur if "Move base model to CPU when using refiner" is enabled with no other options, or only options from the first section, enabled.
